docs/recipes/travis_build_stages.md

```diff
@@ -0,0 +1,90 @@
+# Using semantic-release with [Travis build stages](https://docs.travis-ci.com/user/build-stages/)
+
+![Travis Build Stages screenshot](https://cloud.githubusercontent.com/assets/3729517/25229553/0868909c-25d1-11e7-9263-b076fdef9288.gif)
+
+Build stages are a great way to run semantic release compared to running it in `after_success`:
+
+* The release stage can use Node 8, while the other stages can use other versions, even use a completely different language.
+* It allows you to easily release a project that is not even written in JavaScript, without impacting the job configuration of your test jobs in any way.
+* Instead of having to figure out a "build leader" to do the release and poll the Travis API to wait until all jobs passed, Travis can handle only running the release step after the other jobs passed.
+* If the release failed, you can see at a glance that only the release stage failed while your test stage succeded.
+* Travis can take care of not running the whole release stage at all on non-master branches and PRs.
+
+### Annotated example `.travis.yml`
+
+```yml
+# This could even be a different language
+language: node_js
+
+# Define any version matrix you want for testing
+node_js:
+  - '8'
+  - '6'
+
+# This is inherited by all jobs
+cache:
+  directories:
+    - ~/.npm
+
+# This will be inherited in the release stage
+install:
+  - npm install
+
+# This is the script that will only run for your test stage
+script:
+  - npm run build
```

Comment:
Can you instead recommend to build in a `pretest` and in a `prepublishOnly` script?
This way that avoid forgetting the build step, both on the CI and for contributor who fork the repo.

Reply:
Imo that is not really feasible for any medium-sized project and not the convention at least from what I've seen in build setups (TS, webpack, Babel, ...). When developing, you want to use the watch mode of your build pipeline. Running _another_ build before every test run is just wasted time (it could be anywhere between a few seconds to over a minute). Travis also collapses each command separately so it's easier to see what command actually failed. I think it would also make the example a bit more complex because I would have to highlight that in package.json.

In general, this is just a recipe and as such you are of course always free to not follow it 100%. All of these are mostly directly copied from real projects.

Reply:
As a side there is something that is quire related to `semantic-release` and that has been ask several time on the issue tracker: Having the version set in package.json before running the build script.
This is because some build script need the version to be set to do something with it.

In such situation the best way to solve it is to have the build script being triggered in `prepublishOnly`.

That might a good candidate for another recipe though

Reply:
Yeah, that would be a valid use case that could fit into a recipe.

Reply:
Ok for putting that in another recipe